packages/client/src/features/GameSetup/GameSetup.tsx

```diff
@@ -0,0 +1,112 @@
+import type { FC } from 'react';
+import { useState } from 'react';
+import { Button, Col, Form, InputNumber, Row, Typography } from 'antd';
+import { useIntl } from 'react-intl';
+import { useNavigate } from 'react-router-dom';
+import { useDispatch } from 'react-redux';
+import { setNumberOfPlayers, setPlayers } from 'app/slices/gameSlice';
+import { useForm } from 'react-hook-form';
+import { MAX_NUMBER_OF_PLAYERS, MIN_NUMBER_OF_PLAYERS } from 'constants/main';
+import { messages } from './i18n';
+import { ROUTES } from '../../core/Router';
+import { InputPlayerName } from './InputPlayerName';
+import { InputPlayerColor } from './InputPlayerColor';
+import type { GameSetupFormData } from './types';
+
+const createPlayersArray = (length: Nullable<number>) =>
+  Array.from({ length: length || 0 }).map((_, i) => ({
+    id: `${i}-player`,
+  }));
+
+const isUniqueColors = (colors: string[]) => {
+  const colorsSet = new Set(colors);
+
+  return colorsSet.size === colors.length;
+};
+
+export const GameSetup: FC = () => {
+  const [playersCount, setPlayersCount] = useState(createPlayersArray(2));
+  const dispatch = useDispatch();
+  const navigate = useNavigate();
+  const { formatMessage: fm } = useIntl();
+  const {
+    handleSubmit,
+    formState: { errors },
+    reset,
+    control,
+    getValues,
+    setError,
+    clearErrors,
+  } = useForm<GameSetupFormData>({
+    mode: 'onBlur',
+  });
+
+  const handlePlayersChange = (number: Nullable<number>) => {
+    if (!number) {
+      return;
+    }
+
+    setPlayersCount(createPlayersArray(number));
+  };
+
+  const checkUniquenessColors = (data: GameSetupFormData) => {
+    const colors = Object.entries(data)
+      .filter(([key]) => key.includes('color'))
+      .map(([_, value]) => value);
+
+    const isUniqueness = isUniqueColors(colors);
+
+    isUniqueness
+      ? clearErrors('form')
+      : setError('form', { message: fm(messages.errorColorsUnique) });
+
+    return isUniqueness;
+  };
+
+  const submitHandler = () => {
+    const formData = getValues();
+    const formHasErrors = !checkUniquenessColors(formData);
+
+    if (formHasErrors) {
+      return;
+    }
+
+    dispatch(setNumberOfPlayers(playersCount.length));
+    dispatch(setPlayers(formData));
+    reset();
+    navigate(ROUTES.GAME_PAGE.path);
+  };
```

Comment:
Предлагаю вынести вспомогательные функции checkUniquenessColors, isUniqueColors, createPlayersArray в отдельный модуль, а setError и clearErrors использовать в submitHandler - это разгрузит компонент

```suggestion
  const checkUniquenessColors = (data: GameSetupFormData) => {
    const colors = Object.entries(data)
      .filter(([key]) => key.includes('color'))
      .map(([_, value]) => value);

    return isUniqueColors(colors);
  };

  const submitHandler = () => {
    const formData = getValues();
    const formHasErrors = !checkUniquenessColors(formData);

    if (formHasErrors) {
      setError('form', { message: fm(messages.errorColorsUnique) });
      return;
    }
    
    clearErrors('form');

    dispatch(setNumberOfPlayers(playersCount.length));
    dispatch(setPlayers(formData));
    reset();
    navigate(ROUTES.GAME_PAGE.path);
  };
```